src/lib/storage/segment_access_statistics.hpp

```diff
@@ -0,0 +1,379 @@
+ public:
+  inline static const std::string headers = "Other,IteratorCreate,IteratorSeqAccess,IteratorIncreasingAccess,"
+                                            "IteratorRandomAccess,IteratorDereference,AccessorCreate,AccessorAccess,"
+                                            "DictionaryAccess";
+
+  static SegmentAccessType iterator_access_pattern(const std::shared_ptr<const PosList>& positions);
+
+ private:
+  // There are five states:
+  // 0 (unknown), equivalent to IteratorSeqAccess
+  // 1 (sequentially increasing), difference between two neighboring elements is 0 or 1.
+  // 2 (increasing randomly)
+  // 3 (sequentially decreasing), difference between two neighboring elements is -1 or 0.
+  // 4 (decreasing randomly)
+  // 5 (random access)
+  enum class State {Unknown, SeqInc, RndInc, SeqDec, RndDec, Rnd, Count};
+
+  // There are five possible inputs
+  enum class Input {Zero, One, Positive, NegativeOne, Negative, Count};
+
+  constexpr static const std::array<std::array<State, static_cast<uint32_t>(Input::Count)>, static_cast<uint32_t>(State::Count)>
+    _transitions{{{State::Unknown, State::SeqInc, State::SeqInc, State::SeqDec, State::SeqDec},
+                  {State::SeqInc, State::SeqInc, State::RndInc, State::Rnd, State::Rnd},
+                  {State::RndInc, State::RndInc, State::RndInc, State::Rnd, State::Rnd},
+                  {State::SeqDec, State::Rnd, State::Rnd, State::SeqDec, State::RndDec},
+                  {State::RndDec, State::Rnd, State::Rnd, State::RndDec, State::RndDec},
+                  {State::Rnd, State::Rnd, State::Rnd, State::Rnd, State::Rnd}}};
+};
+// -----------------------------------------------------------------------------------------------------------------
+class AtomicAccessStrategy {
+ public:
+  uint64_t count(SegmentAccessType type) const;
+
+  void reset();
+
+  void increase(SegmentAccessType type, uint64_t count);
+
+  static std::string header();
+
+  std::vector<std::string> to_string() const;
+
+ private:
+  std::array<std::atomic_uint64_t, static_cast<uint32_t>(SegmentAccessType::Count)> _count;
+};
+
+// -----------------------------------------------------------------------------------------------------------------
+class NoAccessStrategy {
+ public:
+  uint64_t count(SegmentAccessType type) const;
+
+  void reset();
+
+  void increase(SegmentAccessType type, uint64_t count);
+
+  static std::string header();
+
+  std::vector<std::string> to_string() const;
+};
+
+// -----------------------------------------------------------------------------------------------------------------
+class AtomicTimedAccessStrategy {
+ public:
+  explicit AtomicTimedAccessStrategy();
+
+  uint64_t count(SegmentAccessType type) const;
+
+  void reset();
+
+  void increase(SegmentAccessType type, uint64_t count);
+
+  static std::string header();
+
+  std::vector<std::string> to_string() const;
+
+  static std::chrono::time_point<std::chrono::steady_clock> start_time;
```

Comment:
We had a discussion before the break on whether the entire timing thing should be the job of the consumer who can retrieve the current, monotonically increasing counter in custom intervals. What happened to that?

Reply:
Yes. WIP. I don't thinh the AtomicTimedAccessStrategy is needed any longer.